### PR TITLE
st2 [1847][IMP] purchase_report_custom_hls: Add fields, hidden some column, show secondary qty

### PR DESCRIPTION
**Pull request by TomohiroAbe-Quartile**
[1847](https://www.quartile.co/web#id=1847&action=771&model=project.task&view_type=form&menu_id=505)

### CODE REVIEW

File: purchase_report_custom_hls/models/purchase_order.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# -*- coding: utf-8 -*-
+
+from odoo import fields, models
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    estimated_time_arrival = fields.Char("Estimated Time of Arrival(ETA)")
+    estimated_time_departure = fields.Char("Estimated Time of Departure(ETD)")
```
Comment (yostashiro) [inline purchase_report_custom_hls/models/purchase_order.py]: I'd suggest changing the fields to `eta_date` and `etd_date`.  ETD/ETA are generic enough.

File: purchase_report_custom_hls/models/res_config_settings.py [hunk 2]
```diff
@@ -11,3 +11,11 @@ class ResConfigSettings(models.TransientModel):
...
         implied_group="purchase_report_custom_hls.group_show_purchase_shipping_address",
     )
+    group_show_purchase_date_planned = fields.Boolean(
+        "Show Request Date in Purchase Report",
+        implied_group="purchase_report_custom_hls.group_show_purchase_date_planned",
+    )
+    group_show_purchase_tax_id = fields.Boolean(
+        "Show Tax in Purchase Report",
+        implied_group="purchase_report_custom_hls.group_show_purchase_tax_id",
+    )
```
Comment (yostashiro) [inline purchase_report_custom_hls/models/res_config_settings.py]: I wouldn't suggest using groups here. Show/hide should rather be selectable per PO (with 'hide' as the default).

File: purchase_report_custom_hls/models/purchase_order.py (hunk 1 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/models/purchase_order.py]: Thank you.
I renamed fields.

File: purchase_report_custom_hls/models/res_config_settings.py (hunk 2 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/models/res_config_settings.py]: Thank you.
I deleted this lines

**Comment by TomohiroAbe-Quartile** [comment]
@yostashiro 
Sorry, I have no idea to hide TAX and Date Req. Columns, and amount_tax field.
Here the current purchase report layout.
purchase_report_layout

**Comment by yostashiro** [comment]
> Sorry, I have no idea to hide TAX and Date Req. Columns, and amount_tax field.

See https://github.com/odoo/odoo/blob/f7938f6be7790f49c5c15991775cb909d57c1303/addons/sale/report/sale_report_templates.xml#L109-L111 for how it's done for the discount field in the quotation print.

File: purchase_report_custom_hls/models/purchase_order.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+from odoo import api, fields, models
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    eta_date = fields.Char("Estimated Time of Arrival(ETA)")
+    etd_date = fields.Char("Estimated Time of Departure(ETD)")
```
Comment (yostashiro) [inline purchase_report_custom_hls/models/purchase_order.py]:
```suggestion
    eta_date = fields.Char("ETA")
    etd_date = fields.Char("ETD")
```

File: purchase_report_custom_hls/models/purchase_order.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    quantity_total = fields.Float("Total Quantity", compute="_compute_quantity_total")
+
+    @api.onchange("order_line")
+    def _compute_quantity_total(self):
+        qty_total = 0
+        for pol in self.order_line:
+            qty_total += pol.product_qty
+        self.quantity_total = qty_total
```
Comment (yostashiro) [inline purchase_report_custom_hls/models/purchase_order.py]:
```suggestion
        for order in self:
            order.quantity_total = sum([x.product_qty for x in order.order_line])
```

File: purchase_report_custom_hls/models/purchase_order.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    eta_date = fields.Char("Estimated Time of Arrival(ETA)")
+    etd_date = fields.Char("Estimated Time of Departure(ETD)")
+    quantity_total = fields.Float("Total Quantity", compute="_compute_quantity_total")
+
+    @api.onchange("order_line")
```
Comment (yostashiro) [inline purchase_report_custom_hls/models/purchase_order.py]:
```suggestion
```
Why onchange?

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: purchase_report_custom_hls/models/purchase_order.py (hunk 5 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/models/purchase_order.py]: Thank you.
I debug and now understand difference between onchange and depends.
Now changed to depends.

File: purchase_report_custom_hls/models/purchase_order.py (hunk 4 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/models/purchase_order.py]: Thank you. Worked as expected.

File: purchase_report_custom_hls/models/purchase_order.py (hunk 3 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/models/purchase_order.py]: Thank you.
Edit description.

File: purchase_report_custom_hls/i18n/ja_JP.po [hunk 6]
```diff
@@ -20,6 +20,26 @@ msgstr ""
 msgid "Config Settings"
 msgstr "コンフィグ設定"
 
+#. module: purchase_report_custom_hls
+#: model:ir.model.fields,field_description:purchase_report_custom_hls.field_purchase_order__eta_date
+msgid "ETA"
+msgstr "到着予定時刻"
```
Comment (yostashiro) [inline purchase_report_custom_hls/i18n/ja_JP.po]:
```suggestion
msgstr "到着予定日"
```

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml [hunk 7]
```diff
@@ -1,19 +1,58 @@
...
+        </xpath>
+        <xpath expr="//span[@t-field='o.amount_tax']"  position="attributes">
+            <attribute name="invisible">1</attribute>
+        </xpath>
+        <xpath expr="//table[@class='table table-sm o_main_table']/thead/tr/th[2]" position="attributes">
+            <attribute name="invisible">1</attribute>
+        </xpath>
+        <xpath expr="//div[@class='col-4']/table[@class='table table-sm']/tr[2]" position="attributes">
+            <attribute name="invisible">1</attribute>
+        </xpath>
```
Comment (TomohiroAbe-Quartile) [inline purchase_report_custom_hls/report/purchase_order_report_templates.xml]: This isn't work as expected.

**Comment by TomohiroAbe-Quartile** [comment]
@yostashiro 
I add new Boolean fields to visible/invisible tax and date_req column.

**Review by kakurai8 — APPROVED** [review]
as a functional review, LGTM

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml [hunk 8]
```diff
@@ -1,19 +1,67 @@
...
+                </div>
             </div>
         </xpath>
+        <xpath expr="//span[@t-field='line.product_uom.name']" position="after">
+            <br/>
+            <span t-field="line.secondary_uom_qty"/>
+            <span t-field="line.secondary_uom_id" groups="uom.group_uom"/>
+        </xpath>
+
+        <xpath expr="//table[@class='table table-sm o_main_table']/thead/tr/th[2]" position="attributes">
```
Comment (yostashiro) [inline purchase_report_custom_hls/report/purchase_order_report_templates.xml]: https://qrtl.slack.com/archives/C011DC4R8J2/p1634476632001100

File: purchase_report_custom_hls/__manifest__.py [hunk 9]
```diff
@@ -8,10 +8,11 @@
     "website": "https://www.quartile.co",
     "license": "LGPL-3",
     "installable": True,
-    "depends": ["purchase_stock", "report_common_hls"],
+    "depends": ["purchase_stock", "report_common_hls", "purchase_order_secondary_unit"],
     "data": [
         "security/purchase_security.xml",
         "report/purchase_order_report_templates.xml",
+        "views/purchase_order_view.xml",
```
Comment (yostashiro) [inline purchase_report_custom_hls/__manifest__.py]:
```suggestion
        "views/purchase_order_views.xml",
```

File: purchase_report_custom_hls/__manifest__.py [hunk 10]
```diff
@@ -8,10 +8,11 @@
     "website": "https://www.quartile.co",
     "license": "LGPL-3",
```
Comment (yostashiro) [inline purchase_report_custom_hls/__manifest__.py]:
```suggestion
    "license": "AGPL-3",
```
For depending on purchase_order_secondary_unit.

File: purchase_report_custom_hls/i18n/ja_JP.po [hunk 11]
```diff
@@ -20,6 +20,26 @@ msgstr ""
...
+
+#. module: purchase_report_custom_hls
+#: model:ir.model.fields,field_description:purchase_report_custom_hls.field_purchase_order__etd_date
+msgid "ETD"
+msgstr "出発予定時刻"
```
Comment (yostashiro) [inline purchase_report_custom_hls/i18n/ja_JP.po]:
```suggestion
msgstr "出荷予定日"
```

File: purchase_report_custom_hls/i18n/ja_JP.po [hunk 12]
```diff
@@ -20,6 +20,26 @@ msgstr ""
...
+
+#. module: purchase_report_custom_hls
+#: model:ir.model,name:purchase_report_custom_hls.model_purchase_order
+msgid "Purchase Order"
+msgstr "購買オーダ"
+
+#. module: purchase_report_custom_hls
+#: model_terms:ir.ui.view,arch_db:purchase_report_custom_hls.report_purchaseorder_document_add_field
+msgid "Quantity Total"
+msgstr "総合計数量"
```
Comment (yostashiro) [inline purchase_report_custom_hls/i18n/ja_JP.po]:
```suggestion
msgstr "合計数量"
```

File: purchase_report_custom_hls/i18n/ja_JP.po [hunk 13]
```diff
@@ -35,3 +55,33 @@ msgstr "購買オーダに出荷先アドレスを表示"
...
+
+#. module: purchase_report_custom_hls
+#: model:ir.model.fields,field_description:purchase_report_custom_hls.field_purchase_order__display_tax
+msgid "Display Tax"
+msgstr "税表示"
+
+#. module: purchase_report_custom_hls
+#: model:ir.model.fields,help:purchase_report_custom_hls.field_purchase_order__display_date_req
+msgid "If checked, display requested date in purchase order report."
+msgstr "チェックすると、予定日が購買オーダに印字されます。"
```
Comment (yostashiro) [inline purchase_report_custom_hls/i18n/ja_JP.po]:
```suggestion
msgstr "選択すると注文書に予定日が印字されます。"
```

File: purchase_report_custom_hls/i18n/ja_JP.po [hunk 14]
```diff
@@ -35,3 +55,33 @@ msgstr "購買オーダに出荷先アドレスを表示"
...
+
+#. module: purchase_report_custom_hls
+#: model:ir.model.fields,help:purchase_report_custom_hls.field_purchase_order__display_tax
+msgid "If checked, display tax in purchase order report."
+msgstr "チェックすると、税が購買オーダに印字されます。"
```
Comment (yostashiro) [inline purchase_report_custom_hls/i18n/ja_JP.po]:
```suggestion
msgstr "選択すると注文書に税が印字されます。"
```

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml [hunk 15]
```diff
@@ -1,19 +1,67 @@
...
+                </div>
+                <div t-if="o.payment_term_id" class="col-3">
+                    <strong>Payment Terms:</strong>
+                    <br />
+                    <span t-field="o.payment_term_id"/>
+                </div>
             </div>
         </xpath>
+        <xpath expr="//span[@t-field='line.product_uom.name']" position="after">
+            <br/>
```
Comment (yostashiro) [inline purchase_report_custom_hls/report/purchase_order_report_templates.xml:34]: Insert `t-if="line.secondary_uom_qty"`.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml [hunk 16]
```diff
@@ -1,19 +1,67 @@
...
+            <span t-field="line.secondary_uom_qty"/>
+            <span t-field="line.secondary_uom_id" groups="uom.group_uom"/>
```
Comment (yostashiro) [inline purchase_report_custom_hls/report/purchase_order_report_templates.xml]:
```suggestion
            <span t-field="line.secondary_uom_id"/>
```
If secondary UOM is available, that should mean group_uom is active.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml [hunk 17]
```diff
@@ -1,19 +1,67 @@
...
+
+        <xpath expr="//span[@t-field='o.amount_tax']/../.."  position="after">
+            <tr>
+                <td>Quantity Total</td>
+                <td class="text-right">
+                    <span t-field="o.quantity_total"/>
+                    <span t-field="o.order_line[0].product_uom"/>
+                </td>
+            </tr>
+        </xpath>
```
Comment (yostashiro) [inline purchase_report_custom_hls/report/purchase_order_report_templates.xml:66]: I personally think it's an odd place to show the total quantities.  Unless the customer specifically requests that the field show the information here, I'd consider putting it in the header part next to the payment terms, perhaps.  We may go as it is for now, but please check with the customer.

File: purchase_report_custom_hls/models/purchase_order.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
+from odoo import api, fields, models
```
Comment (yostashiro) [inline purchase_report_custom_hls/models/purchase_order.py:4]: Add copyright and license.

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: purchase_report_custom_hls/i18n/ja_JP.po (hunk 11 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/i18n/ja_JP.po]: ETA and ETD fields use for negotiation, so I added help and delete translation.
I think it is simple for users to understand.

File: purchase_report_custom_hls/i18n/ja_JP.po (hunk 12 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/i18n/ja_JP.po]: Thank you, I edited it.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml (hunk 15 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml:34]: Thank you.  I added t-tag.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml (hunk 16 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml]: Thank you. I confirmed it works expected.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml (hunk 17 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml:66]: I noted it.

File: purchase_report_custom_hls/models/purchase_order.py (hunk 18 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/models/purchase_order.py:4]: I added. Thank you.

File: purchase_report_custom_hls/__manifest__.py (hunk 10 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/__manifest__.py]: I edited it. Thank you.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml (hunk 8 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml]: Thank you, I replaced.

File: purchase_report_custom_hls/models/purchase_order.py [hunk 19]
```diff
@@ -0,0 +1,30 @@
+# Copyright 2021 Quartile Limited
```
Comment (yostashiro) [inline purchase_report_custom_hls/models/purchase_order.py:1]: Add the license.

File: purchase_report_custom_hls/report/purchase_order_report_templates.xml [hunk 20]
```diff
@@ -1,20 +1,70 @@
...
-        <xpath expr="//div[@class='page']" position="after">
+        <xpath expr="//span[@t-field='line.product_uom.name']" position="after">
+            <br/>
+            <t t-if="line.secondary_uom_qty">
+                <span t-field="line.secondary_uom_qty"/>
+                <span t-field="line.secondary_uom_id" />
+            </t>
+        </xpath>
+
+        <xpath expr="//table[hasclass('table','table-sm','o_main_table')]/thead/tr/th[2]" position="attributes">
```
Comment (yostashiro) [inline purchase_report_custom_hls/report/purchase_order_report_templates.xml]: Get rid of the class elements, only leaving the needed ones, if applicable.

Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/report/purchase_order_report_templates.xml]: I edit that all hasclass has single argument.
It worked as expected.

File: purchase_report_custom_hls/models/purchase_order.py (hunk 19 above)
Comment (TomohiroAbe-Quartile) [reply purchase_report_custom_hls/models/purchase_order.py:1]: Sorry, I added.

**Review by yostashiro — APPROVED** [review]
Code review. LGTM 👍

**Pull request merged**